File: binding/go/rhino_test.go
// Copyright 2021-2023 Picovoice Inc.
//
// You may not use this file except in compliance with the license. A copy of the license is
// located in the "LICENSE" file accompanying this source.
//
// Unless required by applicable law or agreed to in writing, software distributed under the
// License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
// express or implied. See the License for the specific language governing permissions and
// limitations under the License.

package rhino

import (
	"encoding/binary"
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"math"
	"os"
	"path/filepath"
	"reflect"
	"testing"
)

var (
	testAccessKey string
	rhino         Rhino
	withinContextTestParameters []WithinContextTestData
	outOfContextTestParameters  []OutOfContextTestData
)

type WithinContextTestData struct {
	language       string
	context        string
	expectedIntent string
	expectedSlots  map[string]string
}

type OutOfContextTestData struct {
	language string
	context  string
}

func TestMain(m *testing.M) {

	flag.StringVar(&testAccessKey, "access_key", "", "AccessKey for testing")
	flag.Parse()

	withinContextTestParameters, outOfContextTestParameters = loadTestData()
	os.Exit(m.Run())
}

func appendLanguage(s string, language string) string {
	if language == "en" {
		return s
	}
	return s + "_" + language
}

func getTestModelPath(language string) string {
	modelRelPath := fmt.Sprintf(
		"../../lib/common/%s.pv",
		appendLanguage("rhino_params", language))
	modelPath, _ := filepath.Abs(modelRelPath)
	return modelPath
}

func getTestContextPath(language string, context string) string {
	contextRelPath := fmt.Sprintf(
		"../../resources/%s/%s/%s_%s.rhn",
		appendLanguage("contexts", language),
		osName,
		context,
		osName)
	contextPath, _ := filepath.Abs(contextRelPath)
	return contextPath
}

func loadTestData() ([]WithinContextTestData, []OutOfContextTestData) {
	content, err := ioutil.ReadFile("../../resources/test/test_data.json")
	if err != nil {
		log.Fatalf("Could not read test data json: %v", err)
	}

	var testData struct {
		Tests struct {
			WithinContext []struct {
				Language string `json:"language"`
				ContextName  string `json:"context_name"`
				Inference struct {
					Intent string `json:"intent"`
					Slots map[string]string `json:"slots"`
				} `json:"inference"`
			} `json:"within_context"`
			OutOfContext []struct {
				Language string `json:"language"`
				ContextName  string `json:"context_name"`
			} `json:"out_of_context"`
		} `json:"tests"`
	}
	err = json.Unmarshal(content, &testData)
	if err != nil {
		log.Fatalf("Could not decode test data json: %v", err)
	}

	for _, x := range testData.Tests.WithinContext {
		withinContextTestData := WithinContextTestData{
			language:      	x.Language,
			context:       	x.ContextName,
			expectedIntent:	x.Inference.Intent,
			expectedSlots:  x.Inference.Slots,
		}

		withinContextTestParameters = append(withinContextTestParameters, withinContextTestData)
	}

	for _, x := range testData.Tests.OutOfContext {
		outOfContextTestData := OutOfContextTestData{
			language:      	x.Language,
			context:       	x.ContextName,
		}

		outOfContextTestParameters = append(outOfContextTestParameters, outOfContextTestData)
	}
	return withinContextTestParameters, outOfContextTestParameters
}

func runTestCase(t *testing.T, rhino *Rhino, audioFileName string, isWithinContext bool, expectedIntent string, expectedSlots map[string]string) {
	err := rhino.Init()
	if err != nil {
		t.Fatalf("%v", err)
	}
	fmt.Printf("Rhino Version: %s\n", Version)
	fmt.Printf("Frame Length: %d\n", FrameLength)
	fmt.Printf("Sample Rate: %d\n", SampleRate)

	testAudioPath, _ := filepath.Abs(filepath.Join("../../resources/audio_samples", audioFileName))
	data, err := ioutil.ReadFile(testAudioPath)
	if err != nil {
		t.Fatalf("Could not read test file: %v", err)
	}
	data = data[44:] // skip header

	isFinalized := false
	frameLenBytes := FrameLength * 2
	frameCount := int(math.Floor(float64(len(data)) / float64(frameLenBytes)))
	sampleBuffer := make([]int16, FrameLength)
	for i := 0; i < frameCount; i++ {
		start := i * frameLenBytes

		for j := 0; j < FrameLength; j++ {
			dataOffset := start + (j * 2)
			sampleBuffer[j] = int16(binary.LittleEndian.Uint16(data[dataOffset : dataOffset+2]))
		}

		isFinalized, err = rhino.Process(sampleBuffer)
		if err != nil {
			t.Fatalf("Could not read test file: %v", err)
		}

		if isFinalized {
			break
		}
	}

	if !isFinalized {
		t.Fatalf("Rhino reached end of file without finalizing.")
	}

	inference, err := rhino.GetInference()
	if err != nil {
		t.Fatalf("Rhino failed to get inference: \n%v", err)
	}

	if isWithinContext {
		if !inference.IsUnderstood {
			t.Fatalf("Didn't understand.")
		}

		if inference.Intent != expectedIntent {
			t.Fatalf("Incorrect intent '%s'", inference.Intent)
		}

		if !reflect.DeepEqual(inference.Slots, expectedSlots) {
			t.Fatalf("Incorrect slots '%v'", inference.Slots)
		}
	} else {
		if inference.IsUnderstood {
			t.Fatalf("Rhino understood a command outside of its context. %v", inference)
		}
	}
}

func TestWithinContext(t *testing.T) {
	for _, tt := range withinContextTestParameters {
		t.Run(tt.language+" within context", func(t *testing.T) {
			rhino = NewRhino(testAccessKey, getTestContextPath(tt.language, tt.context))
			rhino.ModelPath = getTestModelPath(tt.language)
			runTestCase(
				t,
				&rhino,
				fmt.Sprintf("%s.wav", appendLanguage("test_within_context", tt.language)),
				true,
				tt.expectedIntent,
				tt.expectedSlots)
			delErr := rhino.Delete()
			if delErr != nil {
				t.Fatalf("%v", delErr)
			}
		})
	}
}

func TestOutOfContext(t *testing.T) {
	for _, tt := range outOfContextTestParameters {
		t.Run(tt.language+" out of context", func(t *testing.T) {
			rhino = NewRhino(testAccessKey, getTestContextPath(tt.language, tt.context))
			rhino.ModelPath = getTestModelPath(tt.language)
			runTestCase(
				t,
				&rhino,
				fmt.Sprintf("%s.wav", appendLanguage("test_out_of_context", tt.language)),
				false,
				"",
				nil)
			delErr := rhino.Delete()
			if delErr != nil {
				t.Fatalf("%v", delErr)
			}
		})
	}
}

File: binding/java/test/ai/picovoice/rhino/RhinoTest.java
/*
    Copyright 2018-2023 Picovoice Inc.

    You may not use this file except in compliance with the license. A copy of the license is
    located in the "LICENSE" file accompanying this source.

    Unless required by applicable law or agreed to in writing, software distributed under the
    License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
    express or implied. See the License for the specific language governing permissions and
    limitations under the License.
*/

package ai.picovoice.rhino;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.io.File;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Stream;
import javax.sound.sampled.AudioInputStream;
import javax.sound.sampled.AudioSystem;
import javax.sound.sampled.UnsupportedAudioFileException;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class RhinoTest {

    private final String accessKey = System.getProperty("pvTestingAccessKey");
    private Rhino rhino;

    private static Stream<Arguments> withinContextProvider() throws IOException {
        final JsonObject testDataJson = RhinoTestUtils.loadTestData();
        final JsonArray withinContextData = testDataJson
                .getAsJsonObject("tests")
                .getAsJsonArray("within_context");

        final ArrayList<Arguments> testArgs = new ArrayList<>();
        for (int i = 0; i < withinContextData.size(); i++) {
            final JsonObject testData = withinContextData.get(i).getAsJsonObject();
            final String language = testData.get("language").getAsString();
            final String context = testData.get("context_name").getAsString();
            final String intent = testData.getAsJsonObject("inference").get("intent").getAsString();
            HashMap<String, String> expectedSlotValues = new HashMap<String, String>();
            for (Map.Entry<String, JsonElement> entry : testData.getAsJsonObject("inference").getAsJsonObject("slots").asMap().entrySet()) {
                expectedSlotValues.put(entry.getKey(), entry.getValue().getAsString());
            }
            final String audioFileName = RhinoTestUtils.appendLanguage("test_within_context", language) + ".wav";
            testArgs.add(Arguments.of(
                    language,
                    context,
                    audioFileName,
                    intent,
                    expectedSlotValues)
            );
        }
        return testArgs.stream();
    }

    private static Stream<Arguments> outOfContextProvider() throws IOException {
        final JsonObject testDataJson = RhinoTestUtils.loadTestData();
        final JsonArray outOfContextData = testDataJson
                .getAsJsonObject("tests")
                .getAsJsonArray("out_of_context");

        final ArrayList<Arguments> testArgs = new ArrayList<>();
        for (int i = 0; i < outOfContextData.size(); i++) {
            final JsonObject testData = outOfContextData.get(i).getAsJsonObject();
            final String language = testData.get("language").getAsString();
            final String context = testData.get("context_name").getAsString();
            final String audioFileName = RhinoTestUtils.appendLanguage("test_out_of_context", language) + ".wav";
            testArgs.add(Arguments.of(
                    language,
                    context,
                    audioFileName)
            );
        }
        return testArgs.stream();
    }

    @AfterEach
    void tearDown() {
        rhino.delete();
    }

    @Test
    void getVersion() throws RhinoException {
        rhino = new Rhino.Builder()
                .setAccessKey(accessKey)
                .setContextPath(RhinoTestUtils.getTestContextPath("en", "coffee_maker"))
                .build();
        assertTrue(rhino.getVersion() != null && !rhino.getVersion().equals(""));
    }

    @Test
    void getFrameLength() throws RhinoException {
        rhino = new Rhino.Builder()
                .setAccessKey(accessKey)
                .setContextPath(RhinoTestUtils.getTestContextPath("en", "coffee_maker"))
                .build();
        assertTrue(rhino.getFrameLength() > 0);
    }

    @org.junit.jupiter.api.Test
    void getSampleRate() throws RhinoException {
        rhino = new Rhino.Builder()
                .setAccessKey(accessKey)
                .setContextPath(RhinoTestUtils.getTestContextPath("en", "coffee_maker"))
                .build();
        assertTrue(rhino.getSampleRate() > 0);
    }

    void runTestCase(String audioFileName, boolean isWithinContext, String expectedIntent, Map<String, String> expectedSlots) throws IOException, UnsupportedAudioFileException, RhinoException {
        int frameLen = rhino.getFrameLength();
        String audioFilePath = RhinoTestUtils.getAudioFilePath(audioFileName);
        File testAudioPath = new File(audioFilePath);

        AudioInputStream audioInputStream = AudioSystem.getAudioInputStream(testAudioPath);
        assertEquals(audioInputStream.getFormat().getFrameRate(), 16000);

        int byteDepth = audioInputStream.getFormat().getFrameSize();
        byte[] pcm = new byte[frameLen * byteDepth];
        short[] rhinoFrame = new short[frameLen];

        int numBytesRead = 0;
        boolean isFinalized = false;
        while ((numBytesRead = audioInputStream.read(pcm)) != -1) {
            if (numBytesRead / byteDepth == frameLen) {

                ByteBuffer.wrap(pcm).order(ByteOrder.LITTLE_ENDIAN).asShortBuffer().get(rhinoFrame);
                isFinalized = rhino.process(rhinoFrame);
                if (isFinalized) {
                    break;
                }
            }
        }
        assertTrue(isFinalized);

        RhinoInference inference = rhino.getInference();
        assertEquals(inference.getIsUnderstood(), isWithinContext);

        if (isWithinContext) {
            assertEquals(inference.getIntent(), expectedIntent);
            assertEquals(inference.getSlots(), expectedSlots);
        }
    }

    @ParameterizedTest(name = "testWithinContext for ''{0}''")
    @MethodSource("withinContextProvider")
    void testWithinContext(String language, String context, String audioFileName, String intent, Map<String, String> expectedSlotValues) throws IOException, UnsupportedAudioFileException, RhinoException {
        rhino = new Rhino.Builder()
                .setAccessKey(accessKey)
                .setContextPath(RhinoTestUtils.getTestContextPath(language, context))
                .setModelPath(RhinoTestUtils.getTestModelPath(language))
                .build();

        runTestCase(
                audioFileName,
                true,
                intent,
                expectedSlotValues
        );
    }

    @ParameterizedTest(name = "testOutOfContext for ''{0}''")
    @MethodSource("outOfContextProvider")
    void testOutOfContext(String language, String context, String audioFileName) throws IOException, UnsupportedAudioFileException, RhinoException {
        rhino = new Rhino.Builder()
                .setAccessKey(accessKey)
                .setContextPath(RhinoTestUtils.getTestContextPath(language, context))
                .setModelPath(RhinoTestUtils.getTestModelPath(language))
                .build();

        runTestCase(
                audioFileName,
                false,
                null,
                null
        );
    }
}
